### ISSUE TITLE
504 Gateway Timeout only when generating PDFs that take longer than 30 seconds

### ISSUE THREAD

**Issue by kmollinet** [mileszs/wicked_pdf#835]
## Issue description
I'm attempting to create a PDF for a large report that I'm working on dynamically using wicked pdf. The PDF generates beautifully when I create PDFs that only use a few lines of input data (for example I did one that used 125 lines of input data and it took about 25 seconds to create and looked amazing). However, when I have a lot of data it takes around 105 seconds to query the database and create the report from the report.html.erb file. I do exactly the same process as I do to create the 25 second PDF and I get a 504 gateway timeout error. The processing continues in puma which takes the full 105 seconds, and I get a 200 OK when it finishes. 

I timed how long it takes for the 504 gateway timeout error to show up on the screen, and it takes around 35 seconds each time. I'm wondering if that's the 30 second Heroku timeout for HTTP requests or something else. 

I've attempted to put the render function in a Sidekiq worker class but the rails method "render" is not available there and I can't figure out how to render the pdf in a different way. 

The issue, in short, is that the 504 gateway timeout occurs when I attempt to render the PDF from a controller and I can't figure out how to render it from the sidekiq worker instead because "render" isn't available there. 

## System specifications
All of this is happening on my AWS Cloud9 IDE which is running on an EC2 instance (I think) but also may be running on Heroku somehow. That part baffles me (I inherited this project). 

wicked_pdf gem version (output of `cat Gemfile.lock | grep wicked_pdf`):
kmollinet:~/environment/gradSurvey (master) $ cat Gemfile.lock | grep wicked_pdf
  remote: https://github.com/mileszs/wicked_pdf.git
    wicked_pdf (1.4.0)
  wicked_pdf (~> 1.1)!

wkhtmltopdf version (output of `wkhtmltopdf --version`):
wkhtmltopdf 0.12.4 (with patched qt)

whtmltopdf [provider gem](https://rubygems.org/search?utf8=%E2%9C%93&query=wkhtmltopdf) and version if one is used:
What does this mean???

platform/distribution and version (e.g. Windows 10 / Ubuntu 16.04 / Heroku cedar):
I'm using a Windows 10 computer, but I believe this is running on Ubuntu

Code: 

The code in the controller that takes either 25 seconds for small reports or 105 seconds for large ones and seems to work, but 504s out at times: 
```
            render pdf: pdf_name + " " + pdf_year.to_s,
                disposition: 'attachment',
                page_height: 1300,
                encoding: 'utf8',
                page_size:   'A4',
                footer: {html: {template: 'recent_grad/footer.html.erb'}, spacing: 0 },
                margin:  {   top:    10,                     # default 10 (mm)
                            bottom: 20,
                            left:   10,
                            right:  10 },
                template: "recent_grad/report.html.erb",
                locals: {start: @start, survey: @survey, years: @years, college: @college, department: @department, program: @program, emphasis: @emphasis, questions: @questions}
```
            
The code that is in my sidekiq to try to render the PDF from there instead:

```
Sidekiq.configure_client do |config|
  # config.redis = { db: 1 }
  config.redis = { url: 'redis://172.31.6.51:6379/0' }
end

Sidekiq.configure_server do |config|
  # config.redis = { db: 1 }
  config.redis = { url: 'redis://172.31.6.51:6379/0' }
end  

class PDFWorker
  include Sidekiq::Worker
  sidekiq_options retry: false
  def perform(pdf_name, pdf_year)
    # create an instance of ActionView, so we can use the render method outside of a controller
    av = ::ActionView::Base.new()
    av.view_paths = ::ActionController::Base.view_paths
    av.class_eval do
      include ActionController::UrlWriter
      include ApplicationHelper
    end

    av.render pdf: "mypdf",
      disposition: 'attachment',
      page_height: 1300,
         encoding: 'utf8',
        page_size:   'A4',
           footer: {html: {template: 'recent_grad/footer.html.erb'}, spacing: 0 },
           margin:  {   top:    10,                     # default 10 (mm)
                        bottom: 20,
                        left:   10,
                        right:  10 },
         template: "recent_grad/report.html.erb",
           locals: {start: @start, survey: @survey, years: @years, college: @college, department: @department, program: @program, emphasis: @emphasis, questions: @questions}
  end
end
```

**Comment by unixmonkey** [comment]
In Rails (5? 5.1?), you should be able to use `ApplicationController.render(render_options)` instead of the `class_eval` and `av` trick.

I do think that your host is killing the web request for taking too long, but the Ruby process itself keeps running and finishes, but without benefit to whoever requested the PDF.

Something I'm not seeing in this job is doing something with the PDF once it is created. You could attach it to an email, or use `save_as_pdf: Rails.root.join('public', 'pdfs', 'mypdf.pdf')`.

Please let me know how it goes.

**Comment by kmollinet** [comment]
@unixmonkey thanks so much for the reply! 

My rails version is "Rails 5.1.4"

I attempted what you said, but I'm not sure if I'm doing it right (pretty new to all of this). I took out the ```av``` trick part and put in the ```ApplicationController.render(render_options)``` piece you suggested. Question: is the "render_options" what I have below? 

My code: 

```
# app/workers/pdf_worker.rb
require 'sidekiq'

Sidekiq.configure_client do |config|
  config.redis = { url: 'redis://172.31.6.51:6379/0' }
end

Sidekiq.configure_server do |config|
  config.redis = { url: 'redis://172.31.6.51:6379/0' }
end  

class PDFWorker
  include Sidekiq::Worker
  sidekiq_options retry: false
  def perform(pdf_name, pdf_year)
    ApplicationController.render(pdf: "mypdf",
      disposition: 'attachment',
      page_height: 1300,
         encoding: 'utf8',
        page_size:   'A4',
           footer: {html: {template: 'recent_grad/footer.html.erb'}, spacing: 0 },
           margin:  {   top:    10,                     # default 10 (mm)
                        bottom: 20,
                        left:   10,
                        right:  10 },
         template: "recent_grad/report.html.erb",
           locals: {start: @start, survey: @survey, years: @years, college: @college, department: @department, program: @program, emphasis: @emphasis, questions: @questions}
  )
  end
end
```

The idea would be to

**Comment by kmollinet** [comment]
Accidental close, my bad!

**Comment by unixmonkey** [comment]
I made a little working example app to show the basics of how you might deal with this async process.

The meat of it is in this commit:

https://github.com/unixmonkey/generate_pdf_async_example/commit/85e8dcd98fe6580a8692842fcc5316b931ce4eea

But you should be able to clone this and run it to get a better sense of what's happening.

Let me know if you have any trouble or further questions.

**Comment by kmollinet** [comment]
Wow, thanks so much for the help! I'm implementing what you suggested now (I'll just do exactly what you did to start out with then mix in my actual PDF) and let you know how it goes.

**Comment by kmollinet** [comment]
It almost works! I'm able to output a simple PDF using the method above. The end goal is to be able to view the PDF in browser or download it, how can I do this immediately?

Also, when I render a template with simple html (such as a simple h1 tag) it renders fine. I can also render variables that are created in the HTML template such as the following: 

```
<h1>This is for reals the pdf</h1>
<% @myvar = 2 %>

<%= @myvar %>
<%= @myvar2 %>
```

In this example, "myvar" shows up on the page as "2" as expected, but myvar2 doesn't show up because it was initialized in the controller like this: 

```
    def test_view2
        @myvar2 = 17
    end
```

The reason I experimented with this is because I attempted to render the 26 page long PDF that would be created dynamically from my report.html.erb file, but I got a lot of in-template rails code as text like so:

```
<% if @emphasis %> <% elsif @program %> <% elsif @department %> <% else %> <% end %> <% if @college %> <% else %> <% end %>
<%= @emphasis.emphasis_name + " Emphasis" %>
<%= @program.program_name + " Program" %>
<%= "Department " + @department.department_name %>
<%= @college.college_name %>
BYU
<% if @emphasis %> <% elsif @program %> <% elsif @department %> <% else %> <% end %> <% if @college %> <% else %> <% end %> <% @questions.each do
|question| %> <% if question.question_type.question_type != "Free Response" %> <% if @emphasis %> <% @student_questions =
StudentResponse.joins(:response_option, :student).where('response_options.question_id' => question.id, 'student_responses.survey_id' => @survey.id,
'student_responses.year_id' => params[:years], 'students.emphasis_id' => params[:emphasis_id]) %> <% sq_count =
StudentResponse.check_question_response_num(@student_questions).to_f %> <% elsif @program %> <% @student_questions =
StudentResponse.joins(:response_option, :student).where('response_options.question_id' => question.id, 'student_responses.survey_id' => @survey.id,
'student_responses.year_id' => params[:years], 'students.program_id' => params[:program_id]) %> <% sq_count =
StudentResponse.check_question_response_num(@student_questions).to_f %> <% elsif @department %> <% @student_questions =
StudentResponse.joins(:response_option, student: :program).where('response_options.question_id' => question.id, 'student_responses.survey_id' => @survey.id,
'student_responses.year_id' => params[:years], 'programs.department_id' => params[:department_id]) %> <% sq_count =
StudentResponse.check_question_response_num(@student_questions).to_f %> <% end %> <% if @college %> <% @coll_questions =
StudentResponse.joins(:response_option, student: {program: :department}).where('response_options.question_id' => question.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'departments.college_id' => params[:college_id]) %> <% cq_count =
StudentResponse.check_question_response_num(@coll_questions).to_f %> <% end %> <% @univ_questions = StudentResponse.where(question_id: question.id, year_id: @years, survey_id: @survey.id) %> <% un_count = StudentResponse.check_question_response_num(@univ_questions).to_f %> <% @response_options =
ResponseOption.where(question_id: question.id) %> <% @response_options.each do |ro| %> <% if ro.sequence == 1 || ro.sequence == 0 %> <% else %> <% end %>
<% if @emphasis %> <% @student_responses = StudentResponse.check_response_num(@student_questions.joins(:response_option,
:student).where('response_options.id' => ro.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.emphasis_id' => params[:emphasis_id])) %> <% if sq_count == 0 %> <% else %> <% end %> <% elsif @program %> <% @student_responses =
StudentResponse.check_response_num(@student_questions.joins(:response_option, :student).where('response_options.id' => ro.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.program_id' => params[:program_id])) %> <% if sq_count == 0 %> <% else %> <% end %> <%
elsif @department %> <% @student_responses = @student_questions.joins(:response_option, student: :program).where('response_options.id' => ro.id,
'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'programs.department_id' => params[:department_id]).count %> <% if
sq_count == 0 %> <% else %> <% end %> <% else %> <% end %> <% if @college %> <% @coll_responses =
StudentResponse.check_response_num(@coll_questions.joins(:response_option, student: {program: :department}).where('response_options.id' => ro.id,
'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'departments.college_id' => params[:college_id])) %> <% if cq_count ==
0 %> <% else %> <% end %> <% else %> <% end %> <% @univ_responses = StudentResponse.check_response_num(@univ_questions.where(response_option_id:
ro.id)) %> <% if un_count == 0 %> <% else %> <% end %> <% end %> <% if @emphasis %> <% elsif @program %> <% elsif @department %> <% else %> <% end
%> <% if @college %> <% else %> <% end %> <% end %> <% end %> Question
Text
Response Options Count Percent Count Percent Count Percent
<%=
question.question_text %>
<%=
ro.response_text %>
<%= @student_responses %>
0%
<%=
number_to_percentage((@student_responses/sq_count)
* 100, precision: 0) %>
<%= @student_responses %>
0%
<%=
number_to_percentage((@student_responses/sq_count)
* 100, precision: 0) %>
Total
<%=
sq_count.to_i
%>
100%
<%=
sq_count.to_i
%>
100%
<%= sq_count.to_i
%>
100%
<% @job_free_responses.each do |question| %>
<%= question.question_text %>
<% if @emphasis %> <% @fr_responses = StudentResponse.joins(:response_option, :student).where('response_options.question_id' => question.id,
'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.emphasis_id' => params[:emphasis_id]).distinct.first(15) %>
<% elsif @program %> <% @fr_responses = StudentResponse.joins(:response_option, :student).where('response_options.question_id' => question.id,
'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.program_id' => params[:program_id]).distinct.first(15) %> <%
elsif @department %> <% @fr_responses = StudentResponse.joins(:response_option, student: :program).where('response_options.question_id' => question.id,
'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'programs.department_id' => params[:department_id]).distinct.first(15)
%> <% end %> <% @fr_responses.each do |fr| %>
<%= fr.text_response %>
<% end %> <% end %> <% diff = Time.now - @start %> <% puts "--------------------------------------------------------------" %> <% puts "----------------------------------------------------------
----" %> <% puts "--------------------------------------------------------------" %> <% puts "Time Elapsed:" %> <% puts diff %> <% puts "-------------------------------------------------------------
-" %> <% puts "--------------------------------------------------------------" %> <% puts "--------------------------------------------------------------" %>
```

When I simply render the template in the controller, it works fine, but I need it to happen in a job so that it can take longer than 30 seconds (the original issue). 

Basically everything works except, for whatever reason, the results of every file I save from report.html.erb are like above. I'll comment a chunk of the report.html.erb code so you can see what type of complexity it is relative to normal HTML. 

Do you know why this might be happening?

**Comment by kmollinet** [comment]
Here's a little chunk of my report.html.erb file. I hope this shows enough. Basically it's an html.erb file that queries the database in a loop until it runs out of questions to put on the report. The whole thing works when I render directly from the controller, so I don't think it's the report.html.erb file, but doesn't work when rendering from the job. 

```
<% @response_options = ResponseOption.where(question_id: question.id) %>
            
            <% @response_options.each do |ro| %>
              <tr class="smallFont">
                <% if ro.sequence == 1 || ro.sequence == 0 %>
                  <td style="max-width: 10px; vertical-align: middle;" rowspan=<%= @response_options.length %>><div><%= question.question_text %></div></td>
                <% else %>
                  <!--<td></td>-->
                <% end %>
                <% if ro.sequence == 1 || ro.sequence == 0 %>
                  <td style="padding-top:15px;"><div><%= ro.response_text %></div></td>
                <% else %>
                  <td class="padded"><div><%= ro.response_text %></div></td>
                <% end %>
                
                <% if @emphasis %>  <!--abcdef-->
                    <!--Emphasis Count-->
                    <% @student_responses = StudentResponse.check_response_num(@student_questions.joins(:response_option, :student).where('response_options.id' => ro.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.emphasis_id' => params[:emphasis_id])) %>
                    <td class="text-center"><div><%= @student_responses %></div></td>
                    <!--Emphasis Percent-->
                    <% if sq_count == 0 %>
                      <td class="text-center"><div>0%</div></td>
                    <% else %>
                      <td class="text-center"><div><%= number_to_percentage((@student_responses/sq_count) * 100, precision: 0) %></div></td>
                    <% end %>
                    
                <% elsif @program %>
                    <% @student_responses = StudentResponse.check_response_num(@student_questions.joins(:response_option, :student).where('response_options.id' => ro.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'students.program_id' => params[:program_id])) %>
                    <!--Program Count-->
                    <td class="text-center" style="text-align: center;"><div><%= @student_responses %></div></td>
                    <!--Program Percent-->
                    <% if sq_count == 0 %>
                      <td class="text-center"><div>0%</div></td>
                    <% else %>
                      <td class="text-center"><div><%= number_to_percentage((@student_responses/sq_count) * 100, precision: 0) %></div></td>
                    <% end %>        
                <% elsif @department %>
                    <% @student_responses = @student_questions.joins(:response_option, student: :program).where('response_options.id' => ro.id, 'student_responses.survey_id' => @survey.id, 'student_responses.year_id' => params[:years], 'programs.department_id' => params[:department_id]).count %>
                    <!--Department Count-->
                    <td class="text-center"><div><%= @student_responses %></div></td>
                    <!--Department Percent-->
                    <% if sq_count == 0 %>
                      <td class="text-center"><div>0%</div></td>
                    <% else %>
                      <td class="text-center"><div><%= number_to_percentage((@student_responses/sq_count) * 100, precision: 0) %></div></td>
                    <% end %>
                <% else %>
                    <td class="text-center"><div></div></td>
                    <!--Department Percent-->
                    <td class="text-center"><div></div></td> 
                <% end %>
                
```

**Comment by unixmonkey** [comment]
You need to be able to pass those variables from the `render` method (now called in a Job instead of the Controller) into the template. You can use either the `locals` or the `assigns` options to do that.

[I've added a commit showing how that works here](https://github.com/unixmonkey/generate_pdf_async_example/commit/092fb8133dde379554d75633809392c777baf2e9).

**Comment by kmollinet** [comment]
Awesome, I'll try it out. Thanks!

**Comment by kmollinet** [comment]
I think it's working, but I have a lot of variables so I need to do some refactoring. Is there a way to make it show up on the screen instead of show up only in my file system? Or is there a way for it to download immediately when it's done instead of saving it? I know I could do that in the controller, but haven't been successful thus far in the job.

**Comment by unixmonkey** [comment]
The example I put together does make it show up on the screen.

Here's the workflow for that:
1. `/issue835` endpoint gets hit
2. It checks to see if PDF is already there (it won't be the first time, so skip number 3)
3. If PDF is there, then use `send_file` to serve up the PDF (it will never be there the first time though).
4. if PDF is not there, then kick off Job to generate it, and serve HTML
5. HTML refreshes itself every 5 seconds, sending us back to step 1.

Eventually, the PDF is done processing at step 2, and you get served a PDF.

This is a pretty dumbed-down and inefficient workflow, but it doesn't assume anything about what you have available to you.

An Ajax implementation might look like this:
1. `/issue835` endpoint gets hit
2. It kicks off the Job and immediately loads HTML or redirects somewhere with a generating/loading indicator.
3. That page periodically AJAXs a separate endpoint instead of refreshing (like `/issue835.pdf` or `issue835/download.pdf`), that would load the PDF if it were done (this could use a `setInterval` or check with an incremental backoff algorithm)
4. Once the Ajax call gets a 200 response, it redirects to that endpoint, which downloads the pdf.

A websockets or Pusher.js based implementation might look like this:
1. `/issue835` endpoint gets hit
2. It kicks off the Job and immediately loads HTML or redirects somewhere with a generating/loading indicator.
3. That page opens a websocket that subscribes to a channel for that PDF creation
4. When the job finishes, it sends a message on that websocket channel that says it's done.
5. The HTML page gets the websocket event and knows immediately that it is done and redirects or loads the PDF.

Also, instead of saving to a file on the filesystem, you could save it to an ActiveStorage record, or PaperClip upload, or a binary blob in one of your database tables. Instead of `send_file`, you'd use `send_data`. You'll want to do this if on Heroku or another system with ephemeral storage. This way, instead of checking for the file on disk, you'll check for the existence of a database record.

For sure, this is going to be a lot more moving pieces than simply waiting while the PDF gets generated and serving it on-the-fly, but if it's taking more than 30 seconds and your host is cutting off the connection, that's what is needed, unless you can optimize to get under that time.

An alternative might be using a more native PDF generation library like [Prawn](https://github.com/prawnpdf/prawn) with [Prawn-Rails](https://github.com/cortiz/prawn-rails). The upside is this *should* be a lot faster, the downside is you'll be building out your templates in a very different way, and not using HTML or CSS. If you can do a Prawn-based version that consistently works in less than 30 seconds, that might be a good option for you, but as soon as it goes over, you'll need to adopt an async processing technique like the above.

**Comment by kmollinet** [comment]
It worked! It now does exactly what I need it to do. I actually am using Heroku, but each PDF is only required to be shown to the user when they generate the report. They can then either discard the report or download it if the information looks correct. Thank you so much! I appreciate your help! And I hope this thread can help some other people. I'll link it to my various stack overflow inquiries.

**Comment by unixmonkey** [comment]
Ah. I've found your post [here](https://stackoverflow.com/questions/56266501/how-can-i-use-sidekiq-to-render-large-pdfs-as-a-background-process-to-avoid-hero).

Some notes from what read there:

There really isn't much of a difference between a Job and a Worker. If you are using ActiveJob, it's probably using a Sidekiq worker under the hood.

Mike's suggestion is totally valid. You could generate and save out the HTML to a variable in the controller action, which can save you the work of having to translate your templates to work with the Job. This should be a fast operation, then you can hand off the HTML to the Job/Worker as a string, and still process it with `wkhtmltopf` asynchronously, without using WickedPDF at all (though it might still be helpful for asset helpers).

I am not the creator of WickedPDF, I've just been maintaining it for several years. Mike, however is the author of SideKiq, and a really nice person I've had the pleasure of talking to a few times at the past few RubyConfs. If your business depends on Sidekiq, I would recommend you check out his Sidekiq pro commercial offering.

**Comment by kmollinet** [comment]
Interesting, thanks for clearing that up. Do you know why I couldn't use ApplicationController in the worker then? You're right, they're basically just different folders. I assume it's because I wasn't inheriting from ApplicationJob like this: 

```class CreatePdfJob < ApplicationJob```

right? 

I'll update the post, thanks!

**Comment by unixmonkey** [comment]
> Interesting, thanks for clearing that up. Do you know why I couldn't use ApplicationController in the worker then?

I don't know, but I'm sure you could have gotten it working somehow. It might have been as simple as a missing `require` at the top of the worker to load that class, but the ActiveJob inheritance probably took care of loading the right things for you to use it.

BTW, you should configure ActiveJob to use Sidekiq under-the-hood [as shown in the documentation](https://github.com/mperham/sidekiq/wiki/Active-Job) if it isn't already setup that way in your app.

**Comment by kmollinet** [comment]
You're probably right, it was probably something simple like that. The app is now working completely for this part, so I'll close the issue. I won't be using SideKiq (I hope) anymore, so I may just remove that Gem and use jobs the way you showed above. Thanks so much for the detailed help, it was greatly needed as just the basic ideas of what to do weren't quite enough from Stack Overflow. Many thanks!
